### ISSUE TITLE
Storing the default names of elements

### ISSUE THREAD

**Issue by TMiguelT** [pyparsing/pyparsing#223]
So currently elements have their name stored in the `name` field, which is assigned to elements in their `__init__()`, but can then be overwritten with `setName()`.

The problem with this approach is that we then lose access to the original name, which can be useful as a description of what the element parser. In particular I want access to this for the railroad diagram generator.

So I'm proposing a change that allows us to keep the original names. Here's my current suggestion:
```python
from abc import ABC

class Token(ABC):
    # Other methods
    def __init__():
        self.name = None

    def setName(name):
        self.name = name
    
    def __str__():
        return self.name or self.defaultName

    @abc.abstractmethod
    @property
    def defaultName():
        pass

class Literal(Token):
    def __init__(self, matchString):
        super().__init__()
        self.match = matchString

    def defaultName():
        return '"%s"' % str(self.match)
```

This way, I can use `element.defaultName` in my diagram generator. Thoughts?

**Comment by ptmcg** [comment]
I think the idea is fine, but I would make `name` the property, looking at `self._name or self.default_name`, and `default_name` be just an ordinary attribute, and the `name` setter would assign to `self._name`.

**Comment by ptmcg** [comment]
You got me looking at these str and repr methods, and I found a huge amount of copy/paste code (about 100 lines!) for different implementations of `__str__` in the various classes. I've streamlined them now (and in the process, made the Word class emit a better looking string). I think you can define a property using `self.name if self.name is not None else self.strRepr`. `defaultName` is okay, but what do you think of calling it `docName`, as it is useful as the documentation name for an expression (similar to 'docstring')? But either name works for me.

In the course of defaulting `self.name` to None in `ParserElement.__init__`, I have broken your railroad code, so we are currently failing TravisCI tests. I think you will end up replacing direct references to name to references using the new property.

**Comment by TMiguelT** [comment]
Well, nice job cleaning that up! I'm happy to call it `docName`.

>  This property won't need to be abstract, since all the subclasses should set name and/or strRepr appropriately
I would have thought this means it *should* be an abstract method, since that forces child classes to implement it

> In the course of defaulting self.name to None in ParserElement.__init__, I have broken your railroad code, so we are currently failing TravisCI tests.
That's fine, I'm rewriting it anyway since I worked out some massive flaws in it, so I'll just merge in master and fix it all in one go.

**Comment by ptmcg** [comment]
Since all the subclasses do their own management of `name` and `strRepr`, then this property on the base `ParserElement` class should suffice:

    @property
    def docName(self):
        return self.name if self.name is not None else self.strRepr

When you look at the code, you'll see that the method that needs to be abstract is `_make_str_repr`. I've used the old-style `raise NotImplemented`  indication of abstractness, but mark with `@abstract` if you prefer.

**Comment by ptmcg** [comment]
I think I have a temporary fix to the railroad errors that are failing Travis CI tests, just uncommenting:

    # return str(element)

in get_name(). This should not get in your way too much, and will clean up my Travis reports.

**Comment by ptmcg** [comment]
I added ParserElement.recurse(), if you want to change your railroad code to take advantage of it.

**Comment by TMiguelT** [comment]
Great, thanks.

Incidentally, I noticed maybe an inconsistency on the new model. With some elements (e.g. `Or`), you do *not* set the value of `name` in the constructor, and leave it as `None`, but in other such as `Literal`, you do set it. I would have thought you wanted `name` to be devoted solely to user-specified names, and separated the default names into `_make_str_repr()`. But that isn't the case for many elements. Is that intentional?

**Comment by TMiguelT** [comment]
PPS I don't think this use of `NotImplemented` is correct, if it gets hit, Python says
```
    def _make_str_repr(self):
>       raise NotImplemented
E       TypeError: exceptions must derive from BaseException

pyparsing/core.py:1480: TypeError
```

I believe that [`NotImplemented`](https://docs.python.org/3/library/constants.html#NotImplemented) is only used for certain dunder methods like `__eq__()`. I suspect you should have [`NotImplementedError`](https://docs.python.org/3/library/exceptions.html#NotImplementedError) here instead. Or, even better, use an ABC which should raise an error at "compile type" (ie Class construction time) rather than runtime.

**Comment by ptmcg** [comment]
Ouch, yes that is a gaping hole in things so far. And unfortunately, I really can’t undo these assignments to `name`, too many places reference that attribute. In fact, I’m a little surprised that nothing broke by assigning it to None in the __init__.

This probably bears some more design and overall cleanup, instead of just my knee-jerk code from yesterday (“weeks of coding can save hours of design”).

We need:
- a `name` attribute or property that will default to the internally-generated name, but can be overwritten using `setName`
- a place to preserve the internally-generated name if name is overwritten (possibly the current `strRepr` - which I should rename to something better)
- `__str__` and `__repr__`  methods

To do:
- Review all the places where `name` is assigned, and replace or replicate the value into `strRepr` (or implement `_make_str_repr` for that class to return this string)

Options:
- Make `name` a readonly property and have it do essentially what `__str__` does presently?
- Add `getNames` method that returns a tuple: (internally-generated name, user-assigned name or "")?
- Other ideas?

**Comment by TMiguelT** [comment]
Okay so yes in general I think we should deal with this in a backward-compatible way. This means that, following the old behaviour, `self.name` starts as the auto-generated (auto) name, but also returns the user-defined (custom) name if this is changed. This means that `self.name` should behave the same as `__str__()`.

Thus, I propose the following (simplified) object model:

```python
from abc import ABC, abstractmethod

class ParserElement(ABC):
    def __init__(self):
        #: An instance variable which stores the user-defined custom name for the element
        self.customName = None
        #: An instance variable which stores the auto-generated name for the element
        self.autoName = self.generateDefault()

    @abstractmethod
    def generateDefault(self) -> str:
        """"
        Child classes must define this method, which defines how the `auto_name` is set
        """"
        pass

    def setName(name: str):
        self.custom_name = name

    @property
    def name(self) -> str:
        # This will use a user-defined name if it's available, but otherwise defaults back to the auto-generated name
        return self.customName or self.autoName

    @name.setter
    def name(self, val: str):
        # This maintains backwards compatibility by allowing assigning to `name`. However this might not be needed
        self.setName(val)

    def __str__() -> str:
        # Defer name selection to `name`, since we desire the same behaviour
        return self.name
```

**Comment by TMiguelT** [comment]
Sidenote: it's a bit odd that we had `_make_str_repr()` when the rest of the codebase is camelCase, is it not?

**Comment by ptmcg** [comment]
It's purely an internal method, and I got a nice suggestion the other day about PEP8-izing the names, so I had snake case on the brain.

Your prototype object model looks like it will work. Fix `custom_name` in `setName` to `customName`. And in the `name` getter, please test explicitly for None: `return self.customName if self.customName is not None else self.autoName`. I'm not sure if it really should be allowed to do `expr.setName("")`, but it is currently not prohibited.

**Comment by ptmcg** [comment]
Also, `generateDefault` should probably be `generateDefaultName` just to be clear what default thing we are generating. And `autoName` might be changed to `defaultName`, to better link with the generate method.

**Comment by TMiguelT** [comment]
Agreed on all those points. I've updated my example to include them.

**Comment by ptmcg** [comment]
Let me make these changes this evening (I'm working from home due to COVID, but that means I should be *working*, which unfortunately means "other stuff than pyparsing"), or take a stab at them yourself if you don't want to wait.

**Comment by ptmcg** [comment]
They will be fairly pervasive in the core.py file though, since just about everywhere that assigns the default name to `self.name` will have to change.

**Comment by ptmcg** [comment]
Be aware of places where the default name _has_ to be lazily evaluated, since the contents might not be known at init time (all the And, Or, etc. classes, plus Forward, but maybe some others I'm not thinking of at the moment). Also, there is a trick in Forward where the name is temporarily set to '...' to avoid infinite name generation recursion.

**Comment by TMiguelT** [comment]
Unfortunately I may not get time to do this whole rework, because this is only tangentially related to my job. But we'll see, since this is blocking the diagram generation that I need.

 > Be aware of places where the default name has to be lazily evaluated, since the contents might not be known at init time (all the And, Or, etc. classes, plus Forward, but maybe some others I'm not thinking of at the moment). Also, there is a trick in Forward where the name is temporarily set to '...' to avoid infinite name generation recursion.

I suppose in this case, the `generateDefaultName()` is going to return a placeholder (like "..."), and then later on some other function is going to manually assign to `self.generatedName`? I think that still works under this model. We just say "`generateDefaultName` must always work, even if this instance is missing key fields".

**Comment by ptmcg** [comment]
I think I hit all the corner cases. Now you can use defaultName and customName independently, or just name to follow the expected naming priority.

**Comment by TMiguelT** [comment]
Great job! From looking at the changes I'd say it looks solid. It's passing all the tests is it?

**Comment by ptmcg** [comment]
Yep - https://travis-ci.org/github/pyparsing/pyparsing

**Comment by TMiguelT** [comment]
Well that works for me, I'll merge from master and see if I can finish my PR.

**Comment by ptmcg** [comment]
Released in 3.0.0a2
